prow/jira/jira.go

```diff
@@ -214,11 +263,49 @@ func (jc *client) JiraURL() string {
 	return jc.url
 }
 
+// Used determines whether the client has been used
+func (jc *client) Used() bool {
+	return jc.clientUsed.Used()
+}
+
 type bearerAuthRoundtripper struct {
 	generator BearerAuthGenerator
 	upstream  http.RoundTripper
 }
 
+// WithFields clones the client, keeping the underlying delegate the same but adding
+// fields to the logging context
+func (jc *client) WithFields(fields logrus.Fields) Client {
+	return &client{
+		clientUsed: jc.clientUsed,
+		upstream:   jc.upstream,
+		logger:     jc.logger.WithFields(fields),
+		delegate:   jc.delegate,
+	}
+}
+
+// ForPlugin clones the client, keeping the underlying delegate the same but adding
+// a plugin identifier and log field
+func (jc *client) ForPlugin(plugin string) Client {
+	pluginLogger := jc.logger.WithField("plugin", plugin)
+	retryingClient := retryablehttp.NewClient()
+	usedFlagTransport := &clientUsedTransport{
+		m:        sync.Mutex{},
+		upstream: retryingClient.HTTPClient.Transport,
+	}
+	retryingClient.HTTPClient.Transport = usedFlagTransport
+	retryingClient.Logger = &retryableHTTPLogrusWrapper{log: pluginLogger}
+	// ignore error as url.String() was passed to the delegate
+	jiraClient, err := newJiraClient(jc.url, jc.options, retryingClient)
+	pluginLogger.Error(err)
```

Comment:
nit: this will put the error in a field called `msg` which people might miss, as `pluginLogger.WithError(err).Error("invalid Jira URL")` will put it in a field `err`